### PR TITLE
refactor: introduce `SourceFiles` to REPL and cleanup

### LINKED ISSUE #4045: REPL/--interactive rationalisation
I propose to simplify and rationalise the REPL/`--interactive` as follows (@magnus-madsen, this is what we've been discussing over e-mail, but I've added the suggestion that we introduce a `repl` command to replace `--interactive`).

* Add a new `repl` command which will replace the `--interactive` flag. I.e.:
  ```
  flix somefile.flix --interactive
  ```
  becomes:
  ```
  flix repl somefile.flix
  ```
* `flix repl` with no arguments will either start a REPL with a project loaded if the current directory is a Flix project, or start a REPL with just the standard library loaded if it is not.
* `flix` with no arguments is equivalent to `flix repl`
* Remove the `:watch` command
* Modify the execution of code typed at the `flix>` prompt to just compile the entered code incrementally (not perform a `:refresh` each time).

I suggest that we do this before moving on to #3989, #4042, and #4044.

### PR DESCRIPTION
**Pull request by paulbutcher**
Fixes #4045

### CODE REVIEW

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
Pretty printing of error messages seem to have been lost somewhere.

**Comment by magnus-madsen** [comment]
You can try `1 + true`

**Comment by magnus-madsen** [comment]
Overall structure looks OK, I will take a deeper look a bit later.

**Comment by paulbutcher** [comment]
> Pretty printing of error messages seem to have been lost somewhere.

D'oh, yes, you're right. I'll get onto it.

**Comment by paulbutcher** [comment]
> > Pretty printing of error messages seem to have been lost somewhere.
> 
> D'oh, yes, you're right. I'll get onto it.

Done

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/runtime/shell/SourceFiles.scala [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import java.io.File
+import scala.collection.mutable
+
+/**
+  * Represents the source files loaded by the REPL
+  *
+  * @param source Either a path representing the current directory (if the REPL is started with no arguments)
+  *               or a seqence of files (if the REPL is started with a list of file to load)
+  */
+class SourceFiles(source: Either[Path, Seq[File]]) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/runtime/shell/SourceFiles.scala:32]: Do we want a custom enum? We usually do that:

```
sealed trait InputMode // or better name

case object InputMode {
  case class Project extends InputMode
  case class Paths extends InputMode
}
```

File: main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala [hunk 2]
```diff
@@ -279,33 +229,68 @@ class Shell(initialPaths: List[Path], options: Options) {
...
 
       case Category.Unknown =>
         // The input is not recognized. Output an error message.
         w.println("Error: Input input cannot be parsed.")
     }
   }
 
+  private def clearFragments() = {
+    for(i <- 0 to fragments.length;
+        name = "$" + i)
```
[the remark this answers is not in the document]
Comment (paulbutcher) [reply main/src/ca/uwaterloo/flix/runtime/shell/Shell.scala:244]: I guess I felt it was more self-documenting, but I have no problem with your alternative and will change.

**Comment by magnus-madsen** [comment]
!!!jar

**Pull request merged**